politeiawww/cmd/cmswww_payment/cmswww_payment.go

```diff
@@ -0,0 +1,322 @@
+		var pdErrorReply www.PDErrorReply
+		decoder := json.NewDecoder(r.Body)
+		if err := decoder.Decode(&pdErrorReply); err != nil {
+			return nil, err
+		}
+
+		return nil, www.PDError{
+			HTTPCode:   r.StatusCode,
+			ErrorReply: pdErrorReply,
+		}
+	}
+
+	responseBody := util.ConvertBodyToByteArray(r.Body, false)
+	return responseBody, nil
+}
+
+func _main() error {
+	flag.Parse()
+
+	var err error
+	var network string
+	var rpcPort string
+	if *testnet {
+		network = "testnet3"
+		// Only set to testnet port if no rpc port flag set
+		if rpcPort != pd.DefaultMainnetPort {
+			rpcPort = pd.DefaultTestnetPort
+		}
+	} else {
+		network = "mainnet"
+	}
+
+	dataDir := util.CleanAndExpandPath(*dataDir)
+	dataDir = filepath.Join(dataDir, network)
+
+	dbRootCert := util.CleanAndExpandPath(*dbRootCert)
+	dbCert := util.CleanAndExpandPath(*dbCert)
+	dbKey := util.CleanAndExpandPath(*dbKey)
+
+	// Normalize rpc host:port
+	rpcHost := util.NormalizeAddress(*rpcHost, rpcPort)
+	u, err := url.Parse("https://" + rpcHost)
+	if err != nil {
+		return err
+	}
+	rpcHost = u.String()
+
+	dIdentity, err := identity.LoadPublicIdentity(*dIdentityFile)
+	if err != nil {
+		return err
+	}
+
+	net := filepath.Base(dataDir)
+	cacheDB, err := cachedb.New(cachedb.UserPoliteiawww, *dbHost,
+		net, dbRootCert, dbCert, dbKey)
+	if err != nil {
+		switch err {
+		case cache.ErrNoVersionRecord:
+			err = fmt.Errorf("cache version record not found; " +
+				"start politeiad to setup the cache")
+		case cache.ErrWrongVersion:
+			err = fmt.Errorf("wrong cache version found; " +
+				"restart politeiad to rebuild the cache")
+		}
+		return fmt.Errorf("cachedb new: %v", err)
+	}
+
+	net = filepath.Base(dataDir)
+	cmsDB, err := cmsdb.New(*dbHost, net, dbRootCert, dbCert, dbKey)
+	if err != nil {
+		return err
+	}
+	err = cmsDB.Setup()
+	if err != nil {
+		return fmt.Errorf("cmsdb setup: %v", err)
+	}
+
+	paid, err := cmsDB.PaymentsByStatus(uint(cms.PaymentStatusPaid))
+	if err != nil {
+		return err
+	}
+	for _, payment := range paid {
+		invoice, err := cacheDB.Record(payment.InvoiceToken)
+		if err != nil {
+			fmt.Printf("couldn't get invoice for payment check: %v, %v",
+				payment.InvoiceToken, err)
+			continue
+		}
+		found := false
+		for _, v := range invoice.Metadata {
+			if v.ID == mdstream.IDInvoicePayment {
+				fmt.Printf("payment for invoice %v found!\n", payment.InvoiceToken)
+				// Do we need to do any checks to make sure anything else
+				// is needed?
+				found = true
+				break
+			}
+		}
+		// If no associated payment metadata are found for the given
+		// invoice, create the new metadata and add to the existing invoice.
+		if !found {
+			fmt.Printf("payment for invoice %v NOT found! creating metadata stream for invoice record\n", payment.InvoiceToken)
+			// Create new backend invoice payment metadata
+			c := mdstream.InvoicePayment{
+				Version:        mdstream.VersionInvoicePayment,
+				TxIDs:          payment.TxIDs,
```

Comment:
When I was testing this the mdstream that was created was:

```
{
  "version": 1,
  "txids": "b6d8aa1dc8970cc42ccf2d51589e5c54f222ec16220c943e87b6239c8b5935fc, 45f281daa0a3bc92d97123c002409ba7d79250ee1b1f14836e46236d0dd02dfa, 6855574295243ea8745b8d4b8aa5ba0a7f198db4bddc4a66122049d82a7658fe, 9accbaf9f53880b8a891f93c6e4ba1543edffa9be62c8d0a17a5715753a5b718, 6006b76d1c943765eceb092e1acdda862ebd835adabd1ef9e7b57477a585d3a5, 92adc3370f46fc02f42188692197cbb95359623345aac61b7a121cb4880da250, 99cfc044907cbd1f56d96e58acd9ed42b4e6da96a42753630c69988bc4cf7c55, 7f556c2f34697813ea6c828192c2118c2b3192b42448b9cb54c712e6d8a0c4a2, 5c224bf6c13e05cf49d7c3b5e2a564121c77d6353c75250d30c7421440e2dfc9, 2cffdacd2dbb91a7afac8682b8bcc7eb2190456c40fb63e310b8d607ea6c77a2",
  "timeupdated": 1575486095,
  "amountreceived": 12023054506
}
```

The invoice details were:
```
  "invoice": {
    "status": 6,
    "timestamp": 1575485081,
    "userid": "00bf22a3-07f2-4aa4-83ed-e40ff6432620",
    "username": "",
    "publickey": "4926338ce622ee06423d4ac129a6d82397691dcf344da5ea798ee677f4dd1be4",
    "signature": "c8e6d03c9c0d35f89215d43a99fdc23336427c64553bdd78cb482d39643c5fd7569427eceb6e81e9d8344ac928a2e355f99de9cc1aae6d8e0f99ef20d3581005",
    "file": null,
    "version": "",
    "input": {
      "version": 0,
      "month": 5,
      "year": 2019,
      "exchangerate": 2776,
      "contractorname": "luke",
      "contractorlocation": "usa",
      "contractorcontact": "lukebp",
      "contractorrate": 50000,
      "paymentaddress": "TsoG5VEhcY6Eczy2n2hTiwtr4cXZXUFj2aN",
      "lineitems": [
        {
          "type": 1,
          "domain": "development",
          "subdomain": "subdomain",
          "description": "description",
          "proposaltoken": "",
          "subuserid": "",
          "labor": 60,
          "expenses": 0
        },
        {
          "type": 2,
          "domain": "marketing",
          "subdomain": "subdomain",
          "description": "description",
          "proposaltoken": "",
          "subuserid": "",
          "labor": 0,
          "expenses": 1500
        }
      ]
    },
    "payment": {
      "token": "",
      "address": "",
      "txids": null,
      "timestarted": 0,
      "timelastupdated": 0,
      "amountneeded": 0,
      "amountreceived": 0,
      "status": 0
    },
    "censorshiprecord": {
      "token": "2758af45ce990b13d3de1efebedfe7f96c51ecca748814abb5cbfcd9cee0d310",
      "merkle": "",
      "signature": ""
    }
  }
}
```

The only txid that should have been included was:
`b6d8aa1dc8970cc42ccf2d51589e5c54f222ec16220c943e87b6239c8b5935fc`

Reply:
Hrm, I think I'd consider this to be an issue with the address watcher.  It shouldn't have included those past txids into the db entry.  I believe it is proper that any txids present in the db entry should be moved to the metadata.

Reply:
This should be investigated and fixed though before this tool can be merged so that it can be tested properly.

Reply:
Actually, I screwed up the math on this.  I set the labor rate to $500/hr and it included old txs until the amount was met.

Shouldn't address watcher be checking the invoice timestamp though to make sure it doesn't include previous txs that have been sent to the address?